components/log-ingestor/tests/test_compression_listener.rs

```diff
@@ -0,0 +1,128 @@
+use std::{sync::Arc, time::Duration};
+
+use anyhow::Result;
+use async_trait::async_trait;
+use clp_rust_utils::s3::ObjectMetadata;
+use log_ingestor::compression::{Buffer, BufferSubmitter, DEFAULT_LISTENER_CAPACITY, Listener};
+use tokio::sync::{Mutex, mpsc};
+
+const TEST_OBJECT_SIZE: usize = 1024;
+
+/// A test submitter that stores submitted buffers in memory for inspection.
+struct TestBufferSubmitter {
+    store: Arc<Mutex<Vec<Vec<ObjectMetadata>>>>,
+}
+
+impl TestBufferSubmitter {
+    fn new() -> Self {
+        Self {
+            store: Arc::new(Mutex::new(Vec::new())),
+        }
+    }
+
+    fn shared_store(&self) -> Arc<Mutex<Vec<Vec<ObjectMetadata>>>> {
+        self.store.clone()
+    }
+}
+
+#[async_trait]
+impl BufferSubmitter for TestBufferSubmitter {
+    async fn submit(&self, buffer: &[ObjectMetadata]) -> Result<()> {
+        let store = self.store.clone();
+        let _submitted_results = store.lock().await.push(buffer.to_vec());
+        Ok(())
+    }
+}
+
+/// Sends a list of objects to the listener via the provided sender.
+async fn send_to_listener(objects: Vec<ObjectMetadata>, sender: mpsc::Sender<ObjectMetadata>) {
+    for obj in objects {
+        sender.send(obj).await.unwrap();
+    }
+}
+
+/// Creates a vector of [`ObjectMetadata`] objects for a given bucket. Each object will have a
+/// unique key and a fixed size of [`TEST_OBJECT_SIZE`].
+///
+/// # Returns
+///
+/// A vector of created objects for testing.
+fn create_test_objects(bucket_name: &str, count: usize) -> Vec<ObjectMetadata> {
+    (0..count)
+        .map(|i| ObjectMetadata {
+            bucket: bucket_name.to_string(),
+            key: format!("object-{i}"),
+            size: TEST_OBJECT_SIZE,
+        })
+        .collect()
+}
+
+#[tokio::test]
```

Comment:
how about
```suggestion
#[tokio::test(start_paused = true)]
```
and add `test-util` feature of tokio to Cargo.toml

This way, we don't need to actually wait 4 secs when running unit test.

Reply:
As discussed offline, we will not use this feature for this particular test because `start_paused` only guarantees ordering for `sleep` calls across different coroutines. `Listener`'s implementation is using `select!` in a long-running loop, which complicates the execution model, and we can't find documentations for this type of behavior. We prefer to use real-time testing for now.